### DISCUSSION TITLE
Tpws не помогает решить проблему со смарт тв LG, Samsung (не Android)

### DISCUSSION THREAD

**Discussion by ghost** [bol-van/zapret#234] [Q&A]
^subj. Удалено ибо проблему решать видимо никто не хочет.

**Comment by bol-van** [comment]
tpws --socks
by default it runs in transparent mode and cant be used as proxy
in transparent mode it needs iptables with dnat redirect or tproxy

**Reply by bol-van** [reply]
нельзя использовать прозрачный режим как явный прокси
там совершенно разная логика

**Comment by bol-van** [comment]
--debug  пишет только в консоль
с  --daemon он бесполезен

на счет копирования. можно прямо в опт гит клонировать
копирования не будет. оно сделано чтобы скопировать из tmpfs только минимум
такова особенность опенврт. мало места на диске. гит не влезет

а насчет философии. если сделать инструкцию пошаговой, то это снизит порог входа
но окажется что юзера совершенно беспомощны при любой нештатной ситуации
и специфика вопроса такова, что сделать пошагово не понимая вопроса может не получиться
мне за это не платят чтобы сидеть целый день и отвечать на глупые вопросы

потому я принял решение писать в понятийном варианте
не нажмите вот тут , а сделайте осмысленное действие
поймите зачем и что оно дает

это отсеивает нежелательную для меня аудиторию
кому интересен сам вопрос , с тем мне обычно интересно пообщаться
а с нажимателями нет

**Comment by bol-van** [comment]
про та многие пишут, что не работает. мне не на чем проверить
если тпвс вообще без дурения запустить, тв работает ?

в любом случае требуется снимать дамп трафика и сравнивать
если можете, сделайте
нет, так я сам могу посмотреть дампы. правда это будет не очень скоро

**Reply by ghost** [reply]
Нет под рукой entware - я в соседнем топике уже предлагал помочь, без проблем.
Нет под рукой Smart TV (любого) - я даже не знаю что сказать.. в 2024 году.
Итак по теме:
1. Со стопнутым tpws и без правил редиректа трафика приложение на LG стартует быстро, но само собой видосы не грузит.
2. Со стартанутым tpws и с правилами редиректа трафика, но без "дурения" приложение на LG стартует долго, но стартует.
3. Со стартанутым tpws, с правилами редиректа трафика, и --split-pos=2 приложение на LG не стартует "проверьте интернет (-7)".

Напомню, что у меня только --split-pos=2 стабильно заработал остальные параметры или ни на что не влияют или просто не работают. Ростелеком. Список --hostlist крайне короткий, только домены ютюба и ничего больше.

Насчет tcpdump - есть он у меня, но сформулируйте точнее какой именно трафик вам в дамп записать по каким протоколам и портам, входящий или исходящий и что именно сравнивать/искать, сомневаюсь, что весь трафик локалки вам поможет да и не хотелось бы.

PS Замечено, что обсуждения zapret попали в топ поисковиков и яндекса и гугла, боюсь мы стреляем себе в ногу отрыто обсуждая что работает, а что нет. Надо закрываться и обсуждать хотя бы по почте или в закрытой группе в телеге по одобрению админа. А топики здесь почистить совсем. Ранее вы писали, что вам не нужен поток копипастеров, ну вот он у вас уже есть и дальше больше будет. tcpdump здесь уж точно постить не буду для всех и вся.

**Reply by bol-van** [reply]
да я сам в шоке сколько народу набежало
никому не надо было и тут такое
я никого не приглашал :0

дамп нужен между тв и тпвс
и между тпвс и инетом
еще неплохо бы  tpws --debug

есть средства затирать маки и приватные ип в дампах
не скажу с ходу, надо гуглить

**Reply by ghost** [reply]
Вот я как вебмастер говорю, слишком пропиарились, чистите "хвосты". Пора уже... Хабр на вас часто ссылается и другие репы.
Есть множество решений для ПК/mac и для Андроида на уровне скачал, запустил, работает, но нет решений для локалки в целом, поэтому и к вам идем...
Дамп все ж хотелось бы подробнее услышать чего искать и какие порты в 1 очередь, 443 ?
Я конечно постараюсь, но всеж не моя специализация. Маки чистить вот точно времени нет.
PS СмартТВ не для просмотра 1го канала нужен вовсе, а просто потому что несмарт тв уже нет с картинкой от которой глаза не "вытекают". И тот же ютюб очень удобно и киношки HD-4K, кинескопные тож застал и держал до последнего, но бред это все в сравнении. Как Excel и ручные счеты сравнивать. Свои глаза не казенные.

**Reply by bol-van** [reply]
меня главным образом интересует посмотреть что не так ы режиме tcp proxy
то есть без манипуляций с дуоением
и сравнить с прямым конектом
я старался учесть всякие ситуации типа кто-то бросил конект
чтобы не потерять отосланные данные. буферизацию сделал
но может что-то не учел

потому первый дамп нужен из   lan по фильтру  ip телика. без тпвс
 второй с тпвс без параметров дурения
и паралельно лог  tpws --debug

порт как я понимаю редиректился только 443?

быстро посмотреть не смогу.  не дома

**Reply by ghost** [reply]
Ясно, переписку хотите вести публично)
Ради компа изначально хватило только редиректа tcp 443 на tpws, но пока не настроил редирект tcp 80 и udp 443 на tpws на андройд смартфоне не заработал ютюб.
Т.к. я все-таки вебмастер, а не линукс админ и тем более не embedded linux, то еще раз повторюсь плиз поточнее, что именно искать в tcpdump бегла посмотрев трафик 443 порта я в силу шифрования данных вообще ничего полезного для себя не увидел. Однако, в силу профдеформации наверное, подозреваю, что приложение для смарт тв нифига не понимает шутки юмора с разбивкой описательной части с хедерами на несколько частей меньше чем MTU.
Но больше всего меня напрягает, что даже без параметров вообще tpws все же меняет трафик и не только на сайтах из списка hostlist, некоторые смарты (но далеко не все) особенно китайские такое не понимают (Xiaomi, Poco), а с прошивками на основе Google Pixel например все отлично работает.

Вопрос 1: можете точно сказать ошибаюсь ли я в своих предположениях выше или нет?
и есть ли смысл заруливать на tpws tcp 80 и udp 443.
Вопрос 2: для tcpdump можно пример параметров для вас? IP я само собой свои подставлю и интерфейсы.
ps Про срочность вас никто не торопит, если подробно расскажите где и что искать - думаю разберусь.

**Reply by bol-van** [reply]
там все довольно сложно что смотреть
tcpdump -w /tmp/1.cap -i br-lan host 192.168.1.97
наример так

мне неудобно с тыкалки работать. капы тем более смотреть
можем связаться через приват рутракера
обход блокировок другие способы

**Reply by ghost** [reply]
Дело не в сложности, а думал более узкие рамки будут для tcpdump критичные для анализа дабы лишнего много не цеплялось.
Про рутрекер, хорошо, реанимирую давно заброшенный там акк.

**Reply by ovv77** [reply]
В мануале написано, что для запрета quic (используется на AndroidTV) необходимо поставить запрет на выход наружу UDP 443 к "замедляемым" ресурсам - это может помочь. Во всяком случае мне помогло.

**Reply by ovv77** [reply]
После запрета  UDP 443 - сначала заработало, но через некоторое время - перестало. Приложение пишет нет доступа к сети. Так что поиск решения продолжается.

**Reply by bol-van** [reply]
тема на рутрекере обход блокировок - другие способы - zapret

**Reply by EgorKin** [reply]
DROP пакетов UDP 443 приводит к тому, что офф. клиент YouTube на AndroidTV не может прогрузить интерфейс (так же как и на обычном инете от провайдера без доп правил iptables и tpws). По всей видимости он только через QUIC работает и не переходит на HTTPS в случае недоступности UDP 443 показывая "нет сети" либо полностью серый экран.
Но и REDIRECT его на tpws port тоже не помогает (а tpws вообще умеет работать с QUIC ?)

Если изначально на ТВ раздать интернет через VPN то интерфейс офф клиента загружается и потом переходим на инет с роутера с запущенным tpws и всё работает и проигрывается. До тех пор пока не перезапустишь приложение YouTube. Либо пока ему самому не вздумается рефрешнуть свой интерфейс по окончании просмотра очередного видео.

А так с tpws все работает на всех устройствах сети (из браузеров и на андроид смартфонах с офф. клиентами YouTube) кроме YouTube на AndroidTV. Спасибо за труды Ваши. Надеюсь продолжите разбираться с тем, что нужно для работы интерфейса офф клиента через tpws. nfqws нет возможности запускать т.к. нет нужных модулей на моих роутерах Huawei/Honor (а вот root shell есть)

**Reply by bol-van** [reply]
tpws не работает с UDP.  есть ли действительно quic можно посмотреть в tcpdump
udp port 443

**Reply by EgorKin** [reply]
Спасибо за ответ, дамп попробую сделать через windows ноут (на роутере нет tcpdump).
А отсутствие работы с UDP у tpws - это из-за технической невозможности либо может быть реализовано позднее?
Как я понял добавление в tpws метода обхода fake для UDP 443 решило бы проблемы с AndroidTV.

**Reply by bol-van** [reply]
в прозрачном режиме теоретически можно сделать fake для udp
ipfrag невозможно
но это довольно  много работы и не в приоритете

**Reply by ovv77** [reply]
Не по теме, но временное решение - это установка на AndroidTV Яндекс браузера и просмотр ютуба через него

**Reply by dkipkaev** [reply]
Прошу прощения за глупый вопрос, но если я на OpenWrt меняю что-то в `/opt/zapret/config`, какой следующей командой необходимо подтвердить изменения, перезагрузить что-то?

> В файл /opt/zapret/ipset/zapret-hosts-user.txt следует внести "googlevideo.com", после чего перезапустить zapret.

Вот. А как перезапустить запрет?

**Reply by bol-van** [reply]
по хорошему надо сначала остановить, потом внести изменения и запустить снова
service zapret stop
service zapret start
после обновления листов killall -HIP nfqws tpws  . перезапуск не нужен

**Reply by dkipkaev** [reply]
Ни service, ни systemctl ничего не найдено =) Ладно, пока буду делать reboot. Cпасибо
На компе youtube летает. На WebOS LG SmartTV YouTube не открывается. Даёт ошибку "-7"

**Reply by bol-van** [reply]
service есть в openwrt
но можно и через /etc/init.d/zapret stop|start

**Reply by EgorKin** [reply]
> pws не работает с UDP. есть ли действительно quic можно посмотреть в tcpdump
> udp port 443

снял дамп, UPD 443 и 80 были не заблочены. Интерфейс офф клиента youtube не грузится. (с их DROP через iptables тоже не грузится). запущен tpws --daemon --bind-addr 192.168.1.10 --port 999 --tlsrec=sni --hostspell=HOsT --split-http-req=method --split-pos=2 --hostlist=/etc/tpws-hostlist.txt --pidfile $PIDFILE

на него силами iptables редиректятся tcp 80 и 443

ну и судя по дампу UPD 443 все таки используется:

`19:51:04.086276 IP 192.168.1.219.56423 > arn11s12-in-f10.1e100.net.https: UDP, length 1250
19:51:04.086470 IP 10.255.59.190.56423 > arn11s12-in-f10.1e100.net.https: UDP, length 1250
19:51:04.086568 IP 192.168.1.219.56423 > arn11s12-in-f10.1e100.net.https: UDP, length 1250
19:51:04.086767 IP 192.168.1.219.33446 > arn11s12-in-f10.1e100.net.https: Flags [S], seq 3306626556, win 65535, options [mss 1460,sackOK,TS val 481826279 ecr 0,nop,wscale 6], length 0
19:51:04.086847 IP 192.168.1.219.33446 > arn11s12-in-f10.1e100.net.https: Flags [S], seq 3306626556, win 65535, options [mss 1460,sackOK,TS val 481826279 ecr 0,nop,wscale 6], length 0
19:51:04.087099 IP arn11s12-in-f10.1e100.net.https > 192.168.1.219.33446: Flags [S.], seq 2095017534, ack 3306626557, win 29200, options [mss 1460,nop,nop,sackOK,nop,wscale 0], length 0
19:51:04.087183 IP arn11s12-in-f10.1e100.net.https > 192.168.1.219.33446: Flags [S.], seq 2095017534, ack 3306626557, win 29200, options [mss 1460,nop,nop,sackOK,nop,wscale 0], length 0`

Я конечно совсем не сетевик и может тут как раз видно что по UDP оно не получило ответ и перешло на HTTPS по которому ответ получен?

Вот похоже запрос на проверку соединения
`19:51:04.278192 IP 192.168.1.219.41130 > lr-in-f94.1e100.net.http: Flags [P.], seq 1004941572:1004941760, ack 3307604214, win 1024, length 188: HTTP: HEAD /generate_204 HTTP/1.1
E....T@.@........U.^...P;.1..&..P.......HEAD /generate_204 HTTP/1.1
Host: connectivitycheck.gstatic.com
Connection: keep-alive
User-Agent: CrKey/3.72.425634
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.9`

Только я ответ не вижу, разве что это
`19:51:04.284247 IP arn11s12-in-f10.1e100.net.https > 10.255.59.190.52016: Flags [.], ack 2035, win 268, length 0
E..(....z.__..J.
.;....0^y.MP...P...?m..`

Но попытки сделать запросы по UDP 443 он не бросил:
`19:51:04.401923 IP 192.168.1.219.56423 > arn11s12-in-f10.1e100.net.https: UDP, length 1250`

а вот ответов на UDP 443 запросы я не нашел
хотя к другим доменам по иным портам по UDP обмен идет успешно

Есть кстати что-то про UI youtube (что примечательно они идут по IP6):
`19:51:04.134863 IP6 fe80::9ec5:6f14:9ee5:d28d.40556 > fe80::560d:f9ff:fed0:8ced.domain: 62319+ A? www.youtube.com. (33)
19:51:04.134968 IP6 fe80::9ec5:6f14:9ee5:d28d.40556 > fe80::560d:f9ff:fed0:8ced.domain: 62319+ A? www.youtube.com. (33)
19:51:04.135571 IP 10.255.59.190.45704 > ns1.xn--80aana1acrwhec.xn--p1ai.domain: 19710+ A? www.youtube.com. (33)
19:51:04.135958 IP 10.255.59.190.52016 > arn11s12-in-f10.1e100.net.https: Flags [P.], seq 3:549, ack 1, win 1700, length 546
19:51:04.137116 IP ns1.xn--80aana1acrwhec.xn--p1ai.domain > 10.255.59.190.45704: 19710 3/4/8 CNAME youtube-ui.l.google.com., CNAME wide-youtube.l.google.com., A 142.251.170.198 (358)
19:51:04.137855 IP6 fe80::560d:f9ff:fed0:8ced.domain > fe80::9ec5:6f14:9ee5:d28d.40556: 62319 3/4/8 CNAME youtube-ui.l.google.com., CNAME wide-youtube.l.google.com., A 142.251.170.198 (358)
19:51:04.137985 IP6 fe80::560d:f9ff:fed0:8ced.domain > fe80::9ec5:6f14:9ee5:d28d.40556: 62319 3/4/8 CNAME youtube-ui.l.google.com., CNAME wide-youtube.l.google.com., A 142.251.170.198 (358)
`
кажется это ответы:
`19:51:09.131333 IP6 fe80::9ec5:6f14:9ee5:d28d > fe80::560d:f9ff:fed0:8ced: ICMP6, neighbor solicitation, who has fe80::560d:f9ff:fed0:8ced, length 32
19:51:09.131567 IP6 fe80::9ec5:6f14:9ee5:d28d > fe80::560d:f9ff:fed0:8ced: ICMP6, neighbor solicitation, who has fe80::560d:f9ff:fed0:8ced, length 32
19:51:09.132182 IP6 fe80::560d:f9ff:fed0:8ced > fe80::9ec5:6f14:9ee5:d28d: ICMP6, neighbor advertisement, tgt is fe80::560d:f9ff:fed0:8ced, length 24
19:51:09.132468 IP6 fe80::560d:f9ff:fed0:8ced > fe80::9ec5:6f14:9ee5:d28d: ICMP6, neighbor advertisement, tgt is fe80::560d:f9ff:fed0:8ced, length 24`

В общем взглянуть бы Вам самим на AndroidTV что происходит, а я совсем не спец по сетям.

**Reply by bol-van** [reply]
у меня нет android tv.
чтобы оценить происходящее, нужен дамп с внутреннего и внешнего интерфейса с включенным tpws и лог tpws --debug=2
quic можно заблокировать средствами роутера (не сток).   блок udp:443 на выход

**Reply by EgorKin** [reply]
```
iptables -A FORWARD -p udp --dport 443 -j DROP
iptables -A FORWARD -p udp --dport 80 -j DROP
```
не меняют ситуацию никак - на ТВ интерфейс YouTube также не грузится, на остальных устройствах в сети всё так же грузится и видео воспроизводится.
pktDump.cap дамп могу скинуть на почту, сделан через веб-интерфейс роутера (там есть вкладка для разработчиков с соотв. кнопкой) когда на нем был запущен tpws.` tcpdump -r pktDump.cap` его успешно обрабатывает. Дамп без tpws но с включенным VPN для прогрузки интерфейса YouTube сделать не могу т.к. VPN не на роутере, а через телефон. tpws --debug 2 сейчас нет т.к. он запускается в режиме --daemon.
Если нужен .cap - напишите на почту из моего профиля тут, вышлю в ответном сообщении.

**Reply by dkipkaev** [reply]
UPD: Поставил youtubeUnblock, отлично работает телевизор с параметрами из issue 59 Waujito/youtubeUnblock#59
Asus RT-N11P, OpenWRT 22.03.7. Старенький роутер, уже 10 лет ему, валялся без дела. Вроде попадает под предупреждение 4/32, но работает. 
К Wi-Fi не подключиться, но он и не нужен. Просто от основного роутера стоит прокладкой перед телевизором по схеме: Роутер (основной) — кабель — роутер Asus — кабель — ТВ.
